### PR TITLE
Get transaction status and error from last internal transaction

### LINKED ISSUE #673: Pre-Byzantium transactions are not failed if gas_limit = gas_used
For many transactions on pre-Byzantium Ethereum and ETC chain, I'm seeing normal transactions where gas limit = 21000 and gas used = 21000 show failed in the UI and in Postgres. When they are run through the VM these transactions are in fact successful. 

There are a few options here:
1. Have a default status of success if the gas limit = 21000 until we can run the transactions through the VM. If the gas limit > 21000 and gas_limit == gas_used, these transactions are most likely failed due to `out of gas`. A definitive status can be provided once they are run through the VM.

2. Pre-Byzantium transactions should not be given a final status until we run the transaction back through the VM.

I'm open for any other suggestions that might make sense here.

### Environment

* Operating System: Linux

### Steps to reproduce

1. Start indexing Ethereum Classic.
2. Look for failed transactions that have a gas limit = 21000
3. All of these transactions are failed because the gas_limit = gas_used

### Expected behaviour

The transactions where the gas_limit = 21000 and gas_used = 21000 should have a status of success.

### Actual behaviour

They are being marked as failed

### PR DESCRIPTION
**Pull request by KronicDeth**
Fixes #673

## Test Report

1. Against Ethereum Classic (private nodes)
   1. Run until a transactions with error is found: `transaction = Explorer.Repo.one(from t in Explorer.Chain.Transaction, where: not is_nil(t.error), limit: 1)`
       ```
      iex(6)> transaction.block_number
      6557515
      iex(7)> transaction.hash |> to_string
      "0x2296b765ab869497a8b8d386a8e81db792269bcd12af4d2d1c7a83e5c064a6bc"
      ```
   2. Confirm error is a realistic reason: `transaction.error`
      ```
      iex(8)> transaction.error
      "Bad instruction"
      ```
2. Against POA Core
   1. Run until a transactions with error is found: `transaction = Explorer.Repo.one(from t in Explorer.Chain.Transaction, where: not is_nil(t.error), limit: 1)`
      ```
      iex(7)> transaction.block_number
      4501839
      iex(8)> transaction.hash |> to_string()
      "0xeb90950b417b16366955122e4535ecdadbe16ad65830e6cd99c2349de69cf30e"
      ```
   2. Confirm error is a realistic reason: `transaction.error`:
      ```
      iex(9)> transaction.error
      "Reverted"
      ``` 
    
### Enhancements
* Denormalize last (`max(index)`) internal transaction's `error` to new `error` column on the transaction.  This way, we can display the exact error message and not a simple "Failure".
  * `Explorer.Chain.transaction_to_status` uses `error` to give actual `reason` when a transaction fails:
    * `block_hash: nil, status: nil` -> `:pending` [Unchanged]
    * ` status: nil` -> `:awaiting_internal_transactions` pre-Byzantium/ETC [Added]
    * `status: :ok` -> `:success`[Unchanged]
    * `status: :error, internal_transactions_indexed_at: nil, error: nil` -> `{:error, :awaiting_internal_transactions}` post-Byzantium [Added]
    * `status: :error, error: reason` -> `{:error, reason}` All chains and hard-forks [Added]

### Bug Fixes
* Don't infer status based on `gas_limit == gas_used`, it is an imperfect heuristic, @acravenho has made transactions on ETC where `gas_limit == gas_used` has been a success even though the heuristic would have said it was a failure.
  
### Incompatible Changes
* There is no specific "Out of Gas" status from `BlockScoutWeb.TransactionView.format_status`, but `"Error: Out of Gas"` may be returned when the last internal transaction's error is `"Out of Gas"`, but that is directly from the node's reported format and not normalized as it was before.
* `BlockScoutWeb.Transactionview.status` has been changed to `BlockScoutWeb.TransactionView.status_class` and it now returns the full class name as the old way could not handle tuples returned by `Explorer.Chain.transaction_to_status`.
    * `:pending` -> `"Pending"` [Unchanged]
    * `:awaiting_internal_transactions` -> `"(Awaiting internal transactions for status)"` pre-Byzantium/ETC [Added]
    * `:success` - > `"Success"` [Unchanged]
    * `{:error, :awaiting_internal_transactions}` -> `"Error: (Awaiting internal transactions for reason)"`  post-Byzantium [Added]
    * `"Out of Gas"` [Removed]
    * `"Failure"` [Removed]
    * `{:error, reason}` -> `"Error: %{reason}"` All chains and hard-forks [Added]

## Upgrading

This is a change to the database schema to further enforce when `internal_transactions_indexed_at` and `status` can be set and it adds the `error` field:

1. `mix do ecto.drop, ecto.create, ecto.migrate`
2. `MIX_ENV=test mix do ecto.drop, ecto.create, ecto.migrate`
3. Reset logs: `rm -rf logs`
4. Re-index: `iex -S mix`

### CODE REVIEW

File: apps/block_scout_web/lib/block_scout_web/views/transaction_view.ex [hunk 1]
```diff
@@ -46,10 +46,12 @@ defmodule BlockScoutWeb.TransactionView do
     transaction
     |> Chain.transaction_to_status()
     |> case do
```
Comment (KronicDeth) [inline apps/block_scout_web/lib/block_scout_web/views/transaction_view.ex]: @Stamates, @arthurra let me know if you want a different phrasing for this

File: apps/block_scout_web/lib/block_scout_web/views/transaction_view.ex [hunk 2]
```diff
@@ -94,8 +96,14 @@ defmodule BlockScoutWeb.TransactionView do
     |> Base.encode64()
   end
 
-  def status(transaction) do
-    Chain.transaction_to_status(transaction)
+  def status_class(transaction) do
```
Comment (KronicDeth) [inline apps/block_scout_web/lib/block_scout_web/views/transaction_view.ex]: @Stamates, @arthurra let me know if you want me to do this CSS classes differently

File: apps/block_scout_web/lib/block_scout_web/templates/transaction/_tile.html.eex [hunk 3]
```diff
@@ -1,4 +1,4 @@
-<div class="tile tile-type-<%= type_suffix(@transaction) %> fade-in tile-status--<%= status(@transaction) %>" data-test="<%= type_suffix(@transaction) %>" data-transaction-hash="<%= @transaction.hash %>">
+<div class="tile tile-type-<%= type_suffix(@transaction) %> fade-in <%= status_class(@transaction) %>" data-test="<%= type_suffix(@transaction) %>" data-transaction-hash="<%= @transaction.hash %>">
```
Comment (KronicDeth) [inline apps/block_scout_web/lib/block_scout_web/templates/transaction/_tile.html.eex]: @Stamates, @arthurra let me know if you want me to do this CSS classes differently

File: apps/block_scout_web/assets/css/components/_tile.scss [hunk 4]
```diff
@@ -83,7 +83,7 @@
   }
 
   &-status {
-    &--failed {
+    &--error--reason {
```
Comment (KronicDeth) [inline apps/block_scout_web/assets/css/components/_tile.scss]: @Stamates, @arthurra let me know if you want me to do this CSS classes differently.

File: apps/block_scout_web/assets/css/components/_tile.scss [hunk 5]
```diff
@@ -83,7 +83,7 @@
   }
 
   &-status {
```
Comment (KronicDeth) [inline apps/block_scout_web/assets/css/components/_tile.scss]: Do you want to add a different color for `tile-status--awaiting-internal-transactions` or make it use the same color as `tile-status--error--awaiting-internal-transactions`?

**Comment by KronicDeth** [comment]
@acravenho yeah, me too.  I was surprised so many people would mess up the bytecode since I would have thought the compiler wouldn't do that.  I wonder if it's an Ethereum instruction that isn't supported by ETC and it is a contract deployed to both or if its someone trying to trigger a bug in EVM. 🤔

File: apps/explorer/priv/repo/migrations/20180117221923_create_transactions.exs [hunk 6]
```diff
@@ -82,8 +85,67 @@ defmodule Explorer.Repo.Migrations.CreateTransactions do
     create(
       constraint(
         :transactions,
-        :collated_status,
-        check: "block_hash IS NULL OR status IS NOT NULL"
+        :status,
+        # 0 - NULL
+        # 1 - NOT NULL
+        #
+        # | block_hash | internal_transactions_indexed_at | status | OK | description
```
Comment (alexgaribay) [inline apps/explorer/priv/repo/migrations/20180117221923_create_transactions.exs]: 👍 Truth tables

File: apps/explorer/priv/repo/migrations/20180117221923_create_transactions.exs [hunk 7]
```diff
@@ -82,8 +85,67 @@ defmodule Explorer.Repo.Migrations.CreateTransactions do
...
+        # | 0          | 0                                | 0      | 1  | pending
+        # | 0          | 0                                | 1      | 0  | pending with status
+        # | 0          | 1                                | 0      | 0  | pending with internal transactions
+        # | 0          | 1                                | 1      | 0  | pending with internal transactions and status
+        # | 1          | 0                                | 0      | 1  | pre-byzantium collated transaction without internal transactions
+        # | 1          | 0                                | 1      | 1  | post-byzantium collated transaction without internal transactions
+        # | 1          | 1                                | 0      | 0  | pre-byzantium collated transaction with internal transaction without status
+        # | 1          | 1                                | 1      | 1  | pre- or post-byzantium collated transaction with internal transactions and status
+        #
+        # [Karnaugh map](https://en.wikipedia.org/wiki/Karnaugh_map)
```
Comment (alexgaribay) [inline apps/explorer/priv/repo/migrations/20180117221923_create_transactions.exs]: <3 K-map makes it better

**Pull request merged**